blockchain/pseudohsm/keycache.go

```diff
@@ -0,0 +1,267 @@
+package pseudohsm
+
+import (
+	"bufio"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/bytom/crypto/ed25519/chainkd"
+)
+
+// Minimum amount of time between cache reloads. This limit applies if the platform does
+// not support change notifications. It also applies if the keystore directory does not
+// exist yet, the code will attempt to create a watcher at most this often.
+const minReloadInterval = 2 * time.Second
+
+type keysByFile []XPub
+
+func (s keysByFile) Len() int           { return len(s) }
+func (s keysByFile) Less(i, j int) bool { return s[i].File < s[j].File }
+func (s keysByFile) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }
+
+// AmbiguousKeyError is returned when attempting to unlock
+// an XPub for which more than one file exists.
+type AmbiguousKeyError struct {
+	Pubkey  string
+	Matches []XPub
+}
+
+func (err *AmbiguousKeyError) Error() string {
+	files := ""
+	for i, a := range err.Matches {
+		files += a.File
+		if i < len(err.Matches)-1 {
+			files += ", "
+		}
+	}
+	return fmt.Sprintf("multiple keys match keys (%s)", files)
+}
+
+// keyCache is a live index of all keys in the keystore.
+type keyCache struct {
+	keydir   string
+	watcher  *watcher
+	mu       sync.Mutex
+	all      keysByFile
+	byPubs   map[chainkd.XPub][]XPub
+	throttle *time.Timer
+}
+
+func newKeyCache(keydir string) *keyCache {
+	kc := &keyCache{
+		keydir: keydir,
+		byPubs: make(map[chainkd.XPub][]XPub),
+	}
+	kc.watcher = newWatcher(kc)
+	return kc
+}
+
+func (kc *keyCache) hasKey(xpub chainkd.XPub) bool {
+	kc.maybeReload()
+	kc.mu.Lock()
+	defer kc.mu.Unlock()
+	return len(kc.byPubs[xpub]) > 0
+}
+
+func (kc *keyCache) hasAlias(alias string) bool {
+	xpubs := kc.keys()
+	for _, xpub := range xpubs {
+		if xpub.Alias == alias {
+			return true
+		}
+	}
+	return false
+}
+
+func (kc *keyCache) add(newKey XPub) {
+	kc.mu.Lock()
+	defer kc.mu.Unlock()
+
+	i := sort.Search(len(kc.all), func(i int) bool { return kc.all[i].File >= newKey.File })
+	if i < len(kc.all) && kc.all[i] == newKey {
+		return
+	}
+	// newKey is not in the cache.
+	kc.all = append(kc.all, XPub{})
+	copy(kc.all[i+1:], kc.all[i:])
+	kc.all[i] = newKey
+	kc.byPubs[newKey.XPub] = append(kc.byPubs[newKey.XPub], newKey)
+}
+
+func (kc *keyCache) keys() []XPub {
+	kc.maybeReload()
+	kc.mu.Lock()
+	defer kc.mu.Unlock()
+	cpy := make([]XPub, len(kc.all))
+	copy(cpy, kc.all)
+	return cpy
+}
+
+func (kc *keyCache) maybeReload() {
```

Comment:
提出一个疑问：我理解这个func主要的作用就是启动一个kc.watcher来保持cached的更新。 那为什么不在初始化keyCache的时候就把watcher给启动起来，反而要在第一次用到的时候再启动

Reply:
初始化keyCache就启动watcher，会在第一次maybeReload时直接pass， 那么这段时间程序刚启动时可能就无法读取到所有key了

Reply:
我看代码里reload有2处调用， 一处是在 kc.maybeReload()里， 一处是在watch.loop里。 所以逻辑可以变成在初始化kc的时候初始化watch， 在初始化watch的时候调用一次 kc.reload().

不过根据上一个回复解释的原因， 这一个修改只是无关痛痒的优化，可以放到大后期

Reply:
// Wait for file system events and reload.
	// When an event occurs, the reload call is delayed a bit so that
	// multiple events arriving quickly only cause a single reload.  
第二个reload是指钱包目录文件发生变化时，可能触发大量event， 而通过时间延迟可以只触发一次reload